### PR TITLE
Full digilines supporting tp tubes

### PR DESCRIPTION
**Pull request by SwissalpS**
Have done some shallow testing.
I'm open for discussion and suggestions, it was kind of either try and fix tube-tool or add better digilines support to tp-tubes :D

### CODE REVIEW

**Comment by S-S-X** [comment]
Not sure if this is off topic or not but do you mean tube tool still does not work after attempted fix switching to new API?

**Comment by SwissalpS** [comment]
> Not sure if this is off topic or not but do you mean tube tool still does not work after attempted fix switching to new API?

(Yeah, soff-topic)  
I've checked out at b75978c09 and it didn't copy 'can_receive' flag

File: teleport_tube.lua [hunk 1]
```diff
@@ -270,14 +270,32 @@ local function repair_tube(pos, node)
...
+		if nil ~= msg.can_receive then
+			if msg.can_receive == 1 or msg.can_receive == true then
+				can_receive = 1
+			else
+				can_receive = 0
+			end
+		end
+	else
+		return
+	end
```
Comment (OgelGames) [inline teleport_tube.lua:296]:
```suggestion
	if type(msg) == "string" then
		channel = msg
	elseif type(msg) == "table" then
		if type(msg.channel) == "string" then
			channel = msg.channel
		end
		if msg.can_receive == 1 or msg.can_receive == true then
			can_receive = 1
		elseif msg.can_receive == 0 or msg.can_receive == false then
			can_receive = 0
		end
	else
		return
	end
```

**Review by OgelGames — CHANGES_REQUESTED** [review]

File: teleport_tube.lua (hunk 1 above)
Comment (SwissalpS) [reply teleport_tube.lua]: why prefer calling type() multiple times from assigning a var?  
is the reason for readability or for performance?

Comment (S-S-X) [reply teleport_tube.lua]: Probably no real difference with performance but IMO calling type grants better readability and is easier to follow.

For performance: introducing new local variable also has little performance impact, it is not instant operation. Static strings are very well optimized and type should also be very cheap operation.
I wouldn't be surprised if local variable would actually be slower but it does not really matter, in actual game difference probably cannot be even measured.

**Review by S-S-X — APPROVED** [review]
LGTM
Personally I do not like being able to directly program everything in game but can't really come up with any easy interesting ways to do it in this case.

> I've checked out at b75978c09 and it didn't copy 'can_receive' flag

Checked this one, it was actually working but failed to update formspec. If you've used it and wondered why tp tubes do weird things this might be reason (config works but form show previous receive setting). Fixed formspec update issue already so should work with latest version.

**Review by wsor4035 — APPROVED** [review]
seems fine

File: teleport_tube.lua (hunk 1 above)
Comment (OgelGames) [reply teleport_tube.lua]: Yeah, `type()` is such a small function that calling it repeatedly is better for readability than using a variable. And like @S-S-X said, there's practically no performance difference.

**Review by OgelGames — APPROVED** [review]

**Pull request merged**